### ISSUE TITLE
Improvements for the legacy controller

### ISSUE THREAD

**Issue by Rocky04** [WootingKb/wootility-issues#106]
**1 - Changing the range of values from signed to unsigned.**
Some input frameworks (Unity, SDL) aren't comply with the USB HID specification and can't handle controller inputs where the analog range goes from negative to positive values. Due to this they can't accept inputs or they handle negative inputs as positive and with a false center and so it's not usable. For the later an example game is Kerbal Space Program. The emulated DInput for a Xbox 360 controller also has only positive numbers. _While the report for that isn't 100% correct because the upper boundaries misses the leading zero to indicate that the value is not negative._
![Xbox DInput Interface](https://user-images.githubusercontent.com/5910569/127130317-03498670-0ea9-4496-948e-b67513b4c768.png)

**2 - Support the maximum of possible analog axis for DirectInput**
DInput supports only up to 8 analog axis (X, Y, Z, Rx, Ry, Rz, Dial & Slider), where each has an alias from the simulation controls. So if a native axis isn't there but the simulation control alias is resent the axis is supported, but if both are present the simulation control is ignored. Due to this the Wooting support the Slider due to the Throttle alias while the other four aren't working on games which handles the input via DInput. So it would be nice if the two missing native axis are added. More usable analog axis would benefit simulation games.
Here is how the report can look like: https://discord.com/channels/167181566978555904/195657629060694016/552858723513532416

**3 - Adding a point of view hat switch**
The Xbox controller interface has a 8 way hat switch. It would be nice if this is also supported for the legacy type of the Wooting. Right now this is handled as four normal buttons.

**4 - Having the option to choose the output vector type**
Like for the Xbox controller it would be nice if the Fortnite angle and so the circular output vector can be used for the legacy mode, of course as an option. So that the user can toggle if the circular or squared output vector should be used while the squared one is the default value. Some classic gamepads also have a circular input area for their sticks due to the mechanical limitation while some joystick haven't. By this the Wooting can also be used with the special handling for Fortnite on a console (on the XIM Apex).

**5 - Using of a different device ID combination for the controller types**
If each possible analog type state has a different ID Win would be less confused when the analog type is changed. If then the controller interface is moved as the first interface it should not happen anymore that the scheme is shown wrong via the Windows Game Controller panel if the type is none or legacy. Also the legacy controller would be shown in Steam again and so working there while games should no longer detect a controller for the Wooting if it's set to none (there is sometimes still a problem with that). _This would not affect a potential wrong display for the Xbox type but that should not cause as much problems._ 

**6 - Allow to change the starting position for the analog axis**
Right now the starting position or center point for the analog axis are in the middle of their range and for the simulation controllers at the lowest point. It would be nice if there is a way to select where the starting point is. For example it can be done just by choosing how the binding is done. So if both the upper and lower part of an axis is bound the starting is in the middle while if only the lower part is bound the starting is at the lowest point raising to max and if only the upper part is bound the starting is at the max lowering to the bottom if the key is pressed.

**Comment by RphSam** [comment]
Is this done? I am eagerly waiting for this update as i cant use double movement on wooting key board with out it. (I use wooting key board with xim)

**Comment by simon-wh** [comment]
> Is this done? I am eagerly waiting for this update as i cant use double movement on wooting key board with out it. (I use wooting key board with xim)

Not yet, but we intend to include these changes in the next v4 release, do you have a Wooting One/Two or a Lekker/Two HE?

**Comment by RphSam** [comment]
I have w2 and new wooting 2he. I think it will be launched for w 2he first. Any eta for next v4 release?

**Comment by simon-wh** [comment]
> I have w2 and new wooting 2he. I think it will be launched for w 2he first. Any eta for next v4 release?

Ah, cool cool, yeah it'll be launched for W2HE/W2LE first until the W1/W2 support is brought back into v4. ETA wise it's hard to say, Wootility development has been on a bit of a pause the past couple of weeks as some other stuff has come up with higher priority. I should be fully back on Wootility development before the end of this month and hopefully it won't be long until I'll be able to get out the next release including this and DKS improvements

**Comment by RphSam** [comment]
Cool. Thank you for reply. Chers.

**Comment by RphSam** [comment]
Yo Simon! Pleaae dont forget to include fortnite instant straffing on Legacy controller on upcoming update. You are the man. 👍 thanks.

**Comment by simon-wh** [comment]
Just wanted to share an update on this, we've released Wootility v4.3 which has some of the updates to Legacy (now Classic) Controller. Here's the todo list:
- [x] Changing the range of values from signed to unsigned
- [x] Make axis' behave like Joysticks, and by extension be effected by Joystick shape and Fortnite strafing 
- [x] Different Device IDs depending on gamepad mode 
- [ ] Add Dial & Slider axis'
- [ ] Add Hat Switch (Dpad)
- [ ] Allow to change the starting position for the analog axis (this idea still needs some development/thought)

**Comment by mikey-dg** [comment]
When trying to get analog joystick input in older Unreal engine 1/2 games (with a Wooting 2 HE, v2.4.6, Wootility v4.3.4), I've found that normal X/Y/Z axes are handled correctly, but the engine treats X/Y/Z rotation as though they're 100% in one direction, pressing that direction does nothing, and pressing the other direction only returns them to the center point. I assume this is because it's specifically expecting signed input from rotation axes?

It doesn't really matter much for those games, since three axes is plenty for strafing+leaning, and that engine doesn't really handle joystick input well enough for any of it to be practical anyway (can't combine analog/non-analog movement, weird deadzones that eat the first third of input, treating axes more like acceleration vectors than actual axes, etc), but I thought I'd bring it up in case other older games also expect signed input from the rotation axes. Maybe add a checkbox to the Advanced Settings for signed input on rotation axes?

**Comment by Rocky04** [comment]
Are you sure? How can I very it?

I'm not familiar with Unity and so it would take a lot of time just to set it up and get used to program for it. Haven't found an example so far which I can easily download and test for my own to see how it reacts to the inputs.

From the report data it should not handle these axis differently, all six are signed with a center point in the middle. X & Y and Z & Rz are even grouped up together.

**Comment by mikey-dg** [comment]
> Are you sure? How can I very it?
Pretty sure?  When binding these axes in any Unreal 1/2 engine game (I've tried Unreal Tournament, Deus Ex, and UT2004) , they behave as though they're at the maximum value.

> From the report data  ... Unity
Not Unity, Unreal!  And I've not tried this in an Unreal Engine 3+ game, so I have no idea how those will handle it.

**Comment by Rocky04** [comment]
@mikey-dg Oh yeah, mistaken the engine...

I'm not able to reproduce an issue with "Deus Ex: Human Revolution". This game don't even support the classic controller and only works with the Xbox controller and than not in combination with the mouse. I don't have access to another game you mentioned to test these.

_Same with the new upcoming UT which is under development._

I found a demo of UT2004 and there the Wooting in the classic mode didn't worked out of the box but the Xbox mode does. The left stick for the movement worked great but the right stick did nothing while the view was slowly turning a bit to the down left. With another Xbox controller the demo behaved the same. I was able to found [some configurations](https://forums.epicgames.com/unreal-tournament-2003-2004/unreal-tournament-2004-general-chat/117489-how-do-i-configure-my-my-xbox-360-controller?p=1764780#post1764780) which bound the right stick and then it worked and there wasn't any issue. Even the slowly automated turning went away. After changing back to the classic mode with these settings it also worked fine.

**So basically I assume it's more like a configuration problem and you need to check you joystick settings.**

**Comment by mikey-dg** [comment]
Yeah, sorry, I meant the original Deus Ex from 1999, Unreal Engine 1.  I checked settings, and briefly re-checked all three games.

It looks like all three had a look control bound to JoyU, which is the Throttle in classic mode on the Wooting, and a single-key axis.  So that's working as intended, and would explain some of the problems I was seeing.

The rest of the issues are with binding JoyR/V to strafe/movement in the Unreal 1 engine games, which also seems to cause similar behavior (always trying to move in one direction unless you're pressing the opposite axis), but since it only affects movement (and leaning in Deus Ex), and everything works fine when those axes are bound to look, I can only conclude it's a problem with how the game/engine implements movement on those axes.  Sorry to waste your time - I should have tried a few more control bindings to verify that it was the actual controller axis handling and not something odd with the game.

**Comment by Rocky04** [comment]
@mikey-dg So far nearly all axis are working for me. But I don't know how to configure it properly. I'm not even sure if the game supports analog movement but if the axis are applied to the mouse movement everything works fine as expected. So for 5 axis the default value is in the middle but for one it's not.

So to me it looks like Deus Ex support only six controller axis but the Wooting has 7. Due to this it seems Y rotational is not usable in the game because it's dropped.

These are the axis within the game and their mapping to the Wooting:
- JoyX = Wooting X normal
- JoyY = Wooting Y normal
- JoyZ = Wooting Z normal
- JoyR = Wooting Z rotational
- JoyV = Wooting X rotational
- JoyU = Wooting throttle (causing trouble because the base value isn't in the center)
